### ISSUE TITLE
Velocity-family training converges to standing-in-place across all reward/curriculum variants — full diagnostic matrix

### ISSUE THREAD

**Issue by ysaima** [pollen-robotics/microduck_rl#46]
**Title: Velocity-family training converges to standing-in-place across all reward/curriculum variants — full diagnostic matrix**

## Environment

- Repo at `29e887e` (Sep 2, includes `fix not walking` + the velocity2 merge)
- Pinned env from `uv.lock`: warp-lang 1.12.0, mujoco 3.10.0, mjlab 1.3.0, torch 2.9.1+cu128
- 4× L4 (sm_89), driver 570.172.08 (CUDA 12.8), Ubuntu, training on 4096 envs
- Tasks: `Mjlab-Velocity-Flat-MicroDuck`, `Mjlab-VelStand-Flat-MicroDuck`, plus a registered RampReg variant of Velocity-Flat (see below)

## Symptom

Every velocity-family run converges to a **standing / stepping-in-place policy**:

- `Metrics/twist/error_vel_xy` plateaus at ~0.43 (command ranges lin ±0.4 / ang ±1.0, so ≈ mean |cmd|)
- Net displacement 0.1–0.6 m per 20 s episode; `Episode_Reward/air_time ≈ 0.002` (feet never leave the ground) until very late
- Reward breakdown shows `upright` (2.0) and `head_pose_tracking` (2.0) as the dominant positive mass while the robot stands

Meanwhile the **pose tasks train fine** with the same pipeline: StandUp-Flat reaches 90% envs standing within 200 steps; SitStand transitions verified (98% of envs switch posture). So the env/RL plumbing itself works.

## Diagnostic matrix (what we tried)

| # | Hypothesis | Experiment | Result |
|---|---|---|---|
| 1 | Regularizers suppress gait exploration (AGENTS.md "attempt-tax") | Tracking-only rewards (all 11 non-tracking terms weight 0), 1500 iters | error 0.39→0.23, slow but real; still no gait; displacement <1 m |
| 2 | Same + regularizers ramped in at iter 800 (repo's `reward_weight`) | RampReg run, 5000 iters | error dips to 0.15 @1565 then rises to 0.37 — standing wins again |
| 3 | Gait bootstrap missing | air_time (3.0) + foot_swing + foot_clearance at full weight from step 0, all penalties deferred to iter 6000 | error briefly 0.10, air_time activates (0.02), late-episode displacement surge to 0.52 m — then regresses to standing by iter 11300 |
| 4 | Torque insufficiency (BAM M6 force_limit ±1.07 Nm) | vin_range (6.5,8.2)→(6.5,16.4) (force_limit ≈ ×2 via kt·V/R), tracking-only 300 iters | error 0.44 @242 vs 0.29 control — **worse**, hypothesis rejected |
| 5 | DR push_robot events corrupt displacement evals | Disabled push_robot; forced constant forward command via `obs["actor"][48:51]=(0.3,0,0)` | displacement 0.04 m / path 0.11 m over 12 s — policy genuinely does not walk |
| 6 | Official checkpoint as reference | `alpha_walking.onnx` under our sim (privileged obs) | drifts 0.66 m / 20 s — also does not walk (obs-convention mismatch suspected: sensor-view vs privileged, so inconclusive) |
| 7 | Version drift | `uv.lock`-pinned install, verified | ruled out |

Note on #5/#6: we found `push_robot` interval events (3–6 s) firing during train-mode evals — worth flagging for anyone benchmarking displacement on train-mode envs.

## Ask

Our working hypothesis is now that something differs between the HF Jobs training setup and a stock `uv.lock` local run, or that a specific seed/config detail is needed to escape the standing local optimum. Questions:

1. Has anyone hit standing-in-place convergence on the velocity tasks, and what was the escape?
2. Is there a known-good wandb run config (seed, hyperparameters, any uncommitted tweaks) for the merged velocity recipe that we should replicate exactly?
3. For `alpha_walking.onnx`: what observation view does it expect (privileged sim obs vs onboard sensor view), and what normalizer stats/ordering? (Related to #40.)

Happy to share training logs, checkpoints, the RampReg task patch, and the diagnostic scripts (`eval_skill.py`, `disp_eval.py`, `force_cmd_eval.py`).

## What works

StandUp / SitStand / (partially) VelStand-prone-reset recovery — all verified numerically with train-mode evals. The pipeline, DR stack, and eval tooling are solid; it's specifically velocity-gait discovery that never happens.

**Comment by Qi-hub-dot** [comment]
We hit this too on a single RTX 4070 (4096 envs), and I think I can add two data points that aren't in your matrix. Our run converges to error_vel_xy ~= 0.42, so the same place as yours at 0.43.

**1. The air_time gate only looks at the command, not at whether the robot is actually moving.**

```python
# mjlab/tasks/velocity/mdp/rewards.py:L209
def feet_air_time(env, sensor_name, threshold_min=0.05, threshold_max=0.5,
                  command_name=None, command_threshold=0.5):
    ...
    if command_name is not None:
        total_command = torch.norm(command[:, :2], dim=1) + torch.abs(command[:, 2])
        scale = (total_command > command_threshold).float()
        reward *= scale
    return reward
```

`microduck_velocity_env_cfg.py` sets `command_threshold = 0.01`, so any non-zero command passes. A policy that is commanded forward but only steps in place keeps collecting this reward at weight 3.0, which is the largest positive weight in the set.

This may be why your matrix didn't isolate it: you report `air_time ≈ 0.002`, feet never leaving the ground. Ours is `0.9997`, feet leaving constantly but the body not translating. It is a different position in the same basin, and the "step in place and get paid" route is available.

Gating on achieved motion instead (only pay when horizontal speed reaches ~30% of the commanded horizontal speed) took error_vel_xy from 0.43 to 0.35.

**2. The body-frame velocity is dominated by lateral drift, and the tracking reward treats both axes equally.**

Stepping one env with a constant command of `(+0.30, 0.00)` and reading `root_link_lin_vel_b` each step:

| step | forward | lateral |
|---|---|---|
| 3 | +0.044 | -0.056 |
| 5 | +0.072 | -0.149 |
| 7 | +0.112 | **-0.230** |

Lateral drift reaches twice the forward speed. `track_linear_velocity` computes `(cmd_x - v_x)^2 + (cmd_y - v_y)^2`, so it has no reason to prefer fixing the lateral component. Weighting lateral error 3x and raising the tracking weight from 2.0 to 4.0 dropped lateral drift to -0.044 while forward speed stayed the same, and error_vel_xy went to 0.21.

**3. A caution about error_vel_xy as the success signal.**

`velocity_command.py:L64` accumulates

```python
torch.norm(vel_command_b[:, :2] - root_link_lin_vel_b[:, :2], dim=-1) / max_command_step
```

over the whole episode and across all five command modes (standing, heading, world, forward, plain). Measuring the same checkpoint frame by frame over 400 steps gives a steady-state error of **0.07**, while the logged metric reads **0.206**. The metric averages in the ramp-up (error ~0.30 at reset) and the other command modes.

That is roughly a 3x overestimate. It also means the logged number cannot tell you which mode is failing, which is consistent with your matrix being inconclusive on experiments 1 and 3: a run can briefly discover a gait (your 0.15 at iter 1565, our 0.10) and still show a flat 0.4 because the other modes hold the average up.

**Caveats.** Single run per configuration, no seeds, so part of the 0.43 to 0.21 drop could be run-to-run variance. The 30% gate threshold is a guess with no sensitivity analysis. And the steady-state 0.07 is one checkpoint and one reset, though 16 samples across 8 s agree.

Code, diffs and the raw logs are here if useful:
https://github.com/Qi-hub-dot/microduck-locomotion-diagnosis

**Comment by Tim-is-coding** [comment]
A data point from the deployment side, in case it helps: the shipped walking policy shows the same thing in `duck-sim`, and it barely turns either.

Setup: macOS, `microduck` at `daemon-v0.15.0` (a9ec4b2), `microduck_rl` develop at cb70b79 for the body, `scripts/duck-sim` headless, one duck. robotd logs `policy loaded mode="walk" walk=…/policies/current/alpha_walking.onnx`.

From standing, `robot.move` at 10 Hz for 60 s (vx 0.08, vyaw −0.245), reading `robot.state` at 10 Hz:

```
t     policy  move.applied                 odom.position (x, y)   odom.yaw
0 s   stand   [0, 0, 0] (deadman)          0.0401, -0.0112        0.7912
10 s  walk    [0.08, 0.0, -0.245]          0.0399, -0.0086        0.6922
30 s  walk    [0.08, 0.0, -0.245]          0.0399, -0.0089        0.6877
60 s  walk    [0.08, 0.0, -0.245]          0.0398, -0.0090        0.6815
```

`safety.fallen` stays false throughout. So the command arrives and is applied, the duck steps, and the body moves about 2 mm in 60 s where ~4.8 m were commanded. Yaw drops 0.10 rad in the first 10 s (around the stand → walk switch) and then 0.011 rad over the next 50 s, where ~12 rad were commanded.
